File: parser/src/main/java/dev/mvmo/sicklang/parser/ast/expression/HashLiteralExpressionNode.java
package dev.mvmo.sicklang.parser.ast.expression;

import dev.mvmo.sicklang.token.Token;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

import java.util.Map;
import java.util.stream.Collectors;

@Getter
@RequiredArgsConstructor(staticName = "newInstance")
public class HashLiteralExpressionNode implements ExpressionNode {

    private final Token token;
    private final Map<ExpressionNode, ExpressionNode> pairs;

    @Override
    public String tokenLiteral() {
        return token.literal();
    }

    @Override
    public String toString() {
        return String.format("{%s}", pairs.entrySet().stream()
                .map(entry -> entry.getKey().toString() + ":" + entry.getValue().toString())
                .collect(Collectors.joining(", ")));
    }

}

File: parser/src/main/java/dev/mvmo/sicklang/parser/ast/expression/HashLiteralExpressionNode.kt
package dev.mvmo.sicklang.parser.ast.expression

import dev.mvmo.sicklang.token.Token

class HashLiteralExpressionNode(val token: Token, val pairs: Map<ExpressionNode, ExpressionNode>) : ExpressionNode {

    override fun tokenLiteral(): String =
        token.literal

    override fun toString(): String =
        "{${pairs.entries.joinToString { "${it.key}:${it.value}" }}}"

}